src/portable/raspberrypi/rp2040/dcd_rp2040.c

```diff
@@ -247,16 +249,36 @@ static void __tusb_irq_path_func(reset_non_control_endpoints)(void)
 
 static void __tusb_irq_path_func(dcd_rp2040_irq)(void)
 {
+    uint8_t i;
+    uint32_t parked_eps;
     uint32_t const status = usb_hw->ints;
     uint32_t handled = 0;
+    struct hw_endpoint *ep;
 
     if (status & USB_INTF_DEV_SOF_BITS)
     {
       handled |= USB_INTF_DEV_SOF_BITS;
 
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+      parked_eps = (uint32_t)pending_nonperiodic_xfer;
+      last_sof = time_us_32();
+
+      // Kick any deferred nonperiodic IN endpoints
+      for (i = 0; parked_eps && i < USB_MAX_ENDPOINTS; i++) {
+        if (parked_eps & (1u << i)) {
+          parked_eps &= ~(1u << i);
+          ep = hw_endpoint_get_by_num(i, TUSB_DIR_IN);
+          hw_endpoint_lock_update(ep, 1);
+          hw_endpoint_start_next_buffer(ep);
+          hw_endpoint_lock_update(ep, -1);
+        }
+      }
+      pending_nonperiodic_xfer = 0;
```

Comment:
we should only clear what we handled i.e `parked_eps`

```C
pending_nonperiodic_xfer &= ~parked_eps
```

However, since pending_nonperiodic_xfer is accessed by both dcd_edpt_xfer() and sof isr. There is still chance for race condition since both try to read-then-set the variable.

Reply:
Updated to guard the entire section with the endpoint lock. Note that the loop exits if parked_eps = 0, and pending_nonperiodic_xfer is volatile. So taking a local copy avoids emitting tons of loads/stores inside the for loop.

Reply:
Reimplemented in v3